src/util-interval-tree.c

```diff
@@ -0,0 +1,421 @@
+ *        port ranges in the interval tree
+ * \param de_ctx Detection Engine Context
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the root of the tree
+ * \param list A list of DetectPort objects to be filled
+ */
+static void FindOverlaps(DetectEngineCtx *de_ctx, uint16_t port, uint16_t port2,
+        SCIntervalNode *root, DetectPort **list)
+{
+    SCLogDebug("\nfind overlaps for [%u:%u]", port, port2);
+    SCIntervalNode *ptr = root;
+    DetectPort *new_port = NULL;
+    int cnt_overlaps = 0;
+#if 0
+    printIT(de_ctx, root, 1, printIT_SigGroupHeadSigs);
+#endif
+    SCIntervalNode *stack[100], *current = root;
+    memset(&stack, 0, sizeof(stack));
+    int stack_depth = 0;
+    //            stack[stack_depth++] = current;
+
+    //    SCLogDebug("ROOT! [%u:%u]: current:%p", current->port, current->port2, current);
+    while (current || stack_depth) {
+        SCLogDebug("stack_depth %d", stack_depth);
+        for (int x = 0; x < stack_depth && x < 100 && stack[x] != NULL; x++)
+            SCLogDebug("stack[%d]: %p", x, stack[x]);
+
+        while (current != NULL) {
+            SCLogDebug("%s! [%u:%u]: current:%p",
+                    IRB_PARENT(current, irb) == NULL ? "ROOT" : "LEAF", current->port,
+                    current->port2, current);
+            if (current->max < port) {
+                current = NULL;
+                break;
+            }
+
+            ptr = current;
+            const bool is_overlap = IsOverlap(port, port2, ptr);
+            if (is_overlap) {
+                cnt_overlaps++;
+            }
+
+            if (is_overlap && (cnt_overlaps == 1)) {
+                // Allocate memory for port obj only if it's first overlap
+                new_port = DetectPortInit();
+                if (new_port == NULL) {
+                    goto error;
+                }
+
+                SCLogDebug("find overlaps for [%u:%u]: creating new_port for [%u:%u]", port, port2,
+                        port, port2);
+                new_port->port = port;
+                new_port->port2 = port2;
+
+                // PrintSigGroupHeadSigs(de_ctx, ptr->sh);
+                SigGroupHeadCopySigs(de_ctx, ptr->sh, &new_port->sh);
+                // PrintSigGroupHeadSigs(de_ctx, new_port->sh);
+
+                // Since it is guaranteed that the ports received by this stage
+                // will be sorted, insert any new ports to the end of the list
+                // and avoid walking the entire list
+                if (*list == NULL) {
+                    *list = new_port;
+                    (*list)->last = new_port;
+                } else {
+                    DEBUG_VALIDATE_BUG_ON(new_port->port < (*list)->last->port);
+                    (*list)->last->next = new_port;
+                    (*list)->last = new_port;
+                }
+            } else if (new_port != NULL && is_overlap && cnt_overlaps > 1) {
```

Comment:
can cnt_overlaps not be > 1 here? Seems it has to be and thus the check is redundant